core/statedb/statedb.go

```diff
@@ -287,15 +313,74 @@ func (s *StateDB) GetPendingAccount(blockHeight int64) ([]*account.Account, []*a
 		})
 	}
 
+	gasChanged := false
+	for _, delta := range s.StateCache.PendingGasMap {
+		if delta.Cmp(types.ZeroBigInt) > 0 {
+			gasChanged = true
+			break
+		}
+	}
+
+	handledGasAccount := false
 	for index, formatAccount := range s.PendingUpdateAccountMap {
 		if _, exist := s.PendingNewAccountMap[index]; exist {
 			continue
 		}
 
+		if formatAccount.AccountIndex == types.GasAccount && gasChanged {
+			handledGasAccount = true
+			for assetId, delta := range s.StateCache.PendingGasMap {
+				if asset, ok := formatAccount.AssetInfo[assetId]; ok {
+					formatAccount.AssetInfo[assetId].Balance = ffmath.Add(asset.Balance, delta)
+				} else {
+					formatAccount.AssetInfo[assetId] = &types.AccountAsset{
+						Balance:                  delta,
+						OfferCanceledOrFinalized: types.ZeroBigInt,
+					}
+				}
+			}
+		}
+
 		newAccount, err := chain.FromFormatAccountInfo(formatAccount)
 		if err != nil {
 			return nil, nil, nil, err
 		}
+
+		pendingUpdateAccount = append(pendingUpdateAccount, newAccount)
+		pendingNewAccountHistory = append(pendingNewAccountHistory, &account.AccountHistory{
+			AccountIndex:    newAccount.AccountIndex,
+			Nonce:           newAccount.Nonce,
+			CollectionNonce: newAccount.CollectionNonce,
+			AssetInfo:       newAccount.AssetInfo,
+			AssetRoot:       newAccount.AssetRoot,
+			L2BlockHeight:   blockHeight, // TODO: ensure this should be the new block's height.
+		})
+	}
+
+	if !handledGasAccount && gasChanged {
+		gasAccount, err := s.GetAccount(types.GasAccount)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		formatAccount, err := chain.ToFormatAccountInfo(gasAccount)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		for assetId, delta := range s.StateCache.PendingGasMap {
```

Comment:
Can we put the gas account update code into a function, thus it can be reused easily.

Reply:
updated